test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,926 @@
+
+
+Special Handling for preview method, as per JEP 12:
+- When an element is still marked as preview, the `@since` should be the first JDK release where the element was added.
+- If the element is no longer marked as preview, the `@since` should be the first JDK release where it was no longer preview.
+
+note on legacy preview: Until JDK 14, the preview APIs were not marked in any machine-understandable way.
+                        It was deprecated, and had a comment in the javadoc.
+                        and the use of `@PreviewFeature` only became standard in JDK 17.
+                        So the checker has an explicit knowledge of these preview elements.
+
+note: The `<unique-Element-ID>` for methods looks like
+      `method: <erased-return-descriptor> <binary-name-of-enclosing-class>.<method-name>(<ParameterDescriptor>)`.
+it is somewhat inspired from the VM Method Descriptors. But we use the erased return so that methods
+that were later generified remain the same.
+*/
+
+public class SinceChecker {
+    private final Map<String, Set<String>> LEGACY_PREVIEW_METHODS = new HashMap<>();
+    private final Map<String, IntroducedIn> classDictionary = new HashMap<>();
+    private final JavaCompiler tool;
+    private int errorCount = 0;
+
+    // packages to skip during the test
+    private static final Set<String> EXCLUDE_LIST = Set.of(
+            "java.lang.classfile"
+    );
+
+    public static class IntroducedIn {
+        public String introducedPreview;
+        public String introducedStable;
+    }
+
+    public static void main(String[] args) throws Exception {
+        if (args.length == 0) {
+            throw new IllegalArgumentException("Test module not specified");
+        }
+        SinceChecker sinceCheckerTestHelper = new SinceChecker(args[0]);
+        sinceCheckerTestHelper.checkModule(args[0]);
+    }
+
+    private void error(String message) {
+        System.err.println(message);
+        errorCount++;
+    }
+
+    private SinceChecker(String moduleName) throws IOException {
+        tool = ToolProvider.getSystemJavaCompiler();
+        for (int i = 9; i <= Runtime.version().feature(); i++) {
+            DiagnosticListener<? super JavaFileObject> noErrors = d -> {
+                if (!d.getCode().equals("compiler.err.module.not.found")) {
+                    error(d.getMessage(null));
+                }
+            };
+            JavacTask ct = (JavacTask) tool.getTask(null,
+                    null,
+                    noErrors,
+                    List.of("--add-modules", moduleName, "--release", String.valueOf(i)),
+                    null,
+                    Collections.singletonList(SimpleJavaFileObject.forSource(URI.create("myfo:/Test.java"), "")));
+            ct.analyze();
+
+            String version = String.valueOf(i);
+            Elements elements = ct.getElements();
+            elements.getModuleElement("java.base"); // forces module graph to be instantiated
```

Comment:
@lahodaj I've always considered this kind of line to be a workaround; it would be nice if we could have an RFE so this kind of line was unnecessary

Reply:
Thanks. I've filled:
https://bugs.openjdk.org/browse/JDK-8341342